File: server/src/real-smoke.ts
import 'dotenv/config';
import { buildApp } from './app.js';
import { loadConfig } from './config.js';

const config = loadConfig({ DATABASE_PATH: ':memory:', AI_PROVIDER: 'deepseek', LOG_LEVEL: 'warn' });
if (!config.DEEPSEEK_API_KEY) throw new Error('DEEPSEEK_API_KEY is required for the real-provider smoke test');
const app = await buildApp({ config, logger: true });

async function waitForEvaluation(sessionId: number, headers: Record<string, string>) {
  for (let attempt = 0; attempt < 120; attempt += 1) {
    const response = await app.inject({ method: 'GET', url: `/api/sessions/${sessionId}`, headers });
    const result = response.json().result as {
      score: number;
      uncappedScore: number;
      totalWeight: number;
      scoringVersion: string;
      criteria: Array<{ score: number; weight: number; weightedScore: number }>;
    } | null;
    if (result) return result;
    if (response.json().evaluationStatus === 'failed') throw new Error('Evaluator reported a failed status');
    await new Promise((resolve) => setTimeout(resolve, 1000));
  }
  throw new Error('Evaluator did not finish within 120 seconds');
}

try {
  const authResponse = await app.inject({ method: 'POST', url: '/api/auth/demo', payload: { role: 'student' } });
  if (authResponse.statusCode !== 200) throw new Error(`Demo auth failed: HTTP ${authResponse.statusCode}`);
  const token = authResponse.json().token as string;
  const headers = { authorization: `Bearer ${token}` };
  const casesResponse = await app.inject({ method: 'GET', url: '/api/cases', headers });
  const firstCase = casesResponse.json().cases?.[0] as { id: number } | undefined;
  if (!firstCase) throw new Error('No published seed case is available');
  const startResponse = await app.inject({ method: 'POST', url: '/api/sessions', headers, payload: { caseId: firstCase.id } });
  const sessionId = startResponse.json().session?.id as number | undefined;
  if (!sessionId) throw new Error(`Session start failed: HTTP ${startResponse.statusCode}`);
  const messageResponse = await app.inject({
    method: 'POST', url: `/api/sessions/${sessionId}/messages`, headers,
    payload: { message: 'Hello, my name is Alex, a medical student. Could you tell me what brought you in today?' },
  });
  if (messageResponse.statusCode !== 200 || !messageResponse.body.includes('event: complete')) {
    const code = (() => { try { return messageResponse.json().error?.code; } catch { return undefined; } })();
    throw new Error(`Patient actor smoke failed: HTTP ${messageResponse.statusCode}${code ? ` (${code})` : ''}`);
  }
  const completeResponse = await app.inject({ method: 'POST', url: `/api/sessions/${sessionId}/complete`, headers });
  if (completeResponse.statusCode !== 202) throw new Error(`Evaluator queue failed: HTTP ${completeResponse.statusCode}`);
  const result = await waitForEvaluation(sessionId, headers);
  if (result.criteria.length !== 7) throw new Error(`Expected 7 rubric domains, received ${result.criteria.length}`);
  if (result.totalWeight !== 100) throw new Error(`Expected total rubric weight 100, received ${result.totalWeight}`);
  if (result.criteria.reduce((sum, item) => sum + item.weight, 0) !== 100) {
    throw new Error('Criterion weights do not total 100');
  }
  if (result.criteria.some((item) => !Number.isInteger(item.score) || item.score < 0 || item.score > 3)) {
    throw new Error('Evaluator returned a score outside the integer 0–3 anchors');
  }
  const weightedTotal = Math.round(result.criteria.reduce((sum, item) => sum + item.weightedScore, 0) * 100) / 100;
  if (weightedTotal !== result.uncappedScore) {
    throw new Error(`Weighted contributions ${weightedTotal} do not match uncapped score ${result.uncappedScore}`);
  }
  console.log(JSON.stringify({
    ok: true,
    score: result.score,
    uncappedScore: result.uncappedScore,
    criteriaCount: result.criteria.length,
    totalWeight: result.totalWeight,
    scoringVersion: result.scoringVersion,
  }));
} finally {
  await app.close();
}

File: server/app/real_smoke.py
from __future__ import annotations

import math
from typing import Any

from .real_test_support import (
    RealTestFailure,
    emit_status,
    exercise_case,
    login_student,
    published_cases,
    real_provider_client,
    validate_structured_result,
)

SMOKE_QUESTION = "Hello, my name is Alex, a medical student. Could you tell me what brought you in today?"


def _validate_smoke_scoring(result: dict[str, Any], case_slug: str) -> tuple[int | float, int]:
    score, criteria = validate_structured_result(result, case_slug)
    total_weight = result.get("totalWeight")
    uncapped_score = result.get("uncappedScore")
    if (
        isinstance(total_weight, bool)
        or not isinstance(total_weight, (int, float))
        or not math.isclose(float(total_weight), 100, abs_tol=1e-9)
    ):
        raise RealTestFailure("scoring-weight", case_slug)

    criterion_weight = 0.0
    weighted_total = 0.0
    for item in criteria:
        criterion_score = item.get("score")
        weight = item.get("weight")
        weighted_score = item.get("weightedScore")
        if (
            isinstance(criterion_score, bool)
            or not isinstance(criterion_score, (int, float))
            or not math.isfinite(float(criterion_score))
            or not float(criterion_score).is_integer()
            or not 0 <= float(criterion_score) <= 3
            or isinstance(weight, bool)
            or not isinstance(weight, (int, float))
            or isinstance(weighted_score, bool)
            or not isinstance(weighted_score, (int, float))
        ):
            raise RealTestFailure("scoring-domain", case_slug)
        criterion_weight += float(weight)
        weighted_total += float(weighted_score)
    if not math.isclose(criterion_weight, 100, abs_tol=1e-9):
        raise RealTestFailure("scoring-weight", case_slug)
    if (
        isinstance(uncapped_score, bool)
        or not isinstance(uncapped_score, (int, float))
        or not math.isclose(round(weighted_total + 1e-12, 2), float(uncapped_score), abs_tol=1e-9)
    ):
        raise RealTestFailure("scoring-total", case_slug)
    return score, len(criteria)


def main() -> int:
    case_slug = "setup"
    try:
        with real_provider_client() as client:
            headers = login_student(client)
            clinical_case = published_cases(client, headers)[0]
            case_slug = str(clinical_case["slug"])
            result = exercise_case(client, headers, clinical_case, SMOKE_QUESTION)
            score, criteria_count = _validate_smoke_scoring(result, case_slug)
            emit_status("passed", case_slug, score, criteria_count)
        return 0
    except RealTestFailure as error:
        emit_status(f"failed:{error.stage}", error.case_slug, None, 0)
        return 1
    except Exception:
        emit_status("failed:unexpected", case_slug, None, 0)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())
